### PR TITLE
API: Fix listSSHKeyPairs API when listing all resources (listall=true & projectid=-1)

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR fixes the issue noticed when listing all ssh key pairs including ones owned by projects. Exception:
```
(primateqa) 🐱 > list sshkeypairs listall=true projectid=-1
🙈 Error: (HTTP 530, error code 4250) DB Exception on: com.mysql.cj.jdbc.ClientPreparedStatement: SELECT ssh_keypairs.id, ssh_keypairs.uuid, ssh_keypairs.account_id, ssh_keypairs.domain_id, ssh_keypairs.keypair_name, ssh_keypairs.fingerprint, ssh_keypairs.public_key FROM ssh_keypairs  INNER JOIN account ON ssh_keypairs.account_id=account.id  WH ORDER BY ssh_keypairs.id DESC  LIMIT 0, 500
```

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

#### Post Fix:
```
(localcloud) 🐱 > list sshkeypairs listall=true page=1 pagesize=20 filter=name,
{
  "count": 1,
  "sshkeypair": [
    {
      "name": "ssh"
    }
  ]
}
(localcloud) 🐱 > list sshkeypairs listall=true projectid=-1 page=1 pagesize=20 filter=name,
{
  "count": 2,
  "sshkeypair": [
    {
      "name": "test"
    },
    {
      "name": "ssh"
    }
  ]
}
```

### CODE REVIEW

**Comment by Pearl1594** [comment]
Created a new PR & closed https://github.com/apache/cloudstack/pull/5815

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]
tested ok

**Review by DaanHoogland — APPROVED** [review]

**Review by nvazquez — APPROVED** [review]
LGTM

**Comment by nvazquez** [comment]
@blueorangutan test

File: framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java [hunk 1]
```diff
@@ -1287,13 +1289,20 @@ protected void addJoins(StringBuilder str, Collection<JoinBuilder<SearchCriteria
             .append(" ");
             str.insert(fromIndex, onClause);
             String whereClause = join.getT().getWhereClause();
-            if ((whereClause != null) && !"".equals(whereClause)) {
+            if (!StringUtils.isEmpty(whereClause)) {
```
Comment (GutoVeronezi) [inline framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java:1292]:
```suggestion
            if (StringUtils.isNotEmpty(whereClause)) {
```

File: framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java [hunk 2]
```diff
@@ -1287,13 +1289,20 @@ protected void addJoins(StringBuilder str, Collection<JoinBuilder<SearchCriteria
...
+                }
                 str.append(" (").append(whereClause).append(") AND");
             }
             fromIndex += onClause.length();
         }
 
-        str.delete(str.length() - 4, str.length());
+        if (hasWhereClause) {
+            // Remove ' AND' OR 'AND ' at the end of line
+            str.delete(str.length() - 4, str.length());
```
Comment (GutoVeronezi) [inline framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java:1304]: I think it could be a method with an intuitive name instead of a comment.

**Comment by sureshanaparti** [comment]
@Pearl1594 Can you address the outstanding comments. Thanks.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
mannually tested
LGTM

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
retested with the latest changes :+1:

**Review by sureshanaparti — APPROVED** [review]

**Pull request merged**